core/primitives/src/cpu.rs

```diff
@@ -0,0 +1,97 @@
+/// Try to ensure that the required CPU features are available and enabled for the current CPU.
+///
+/// Note that this is a best-effort kind of thing. If these features are already enabled when
+/// compiling the program, then we may very well end up executing some unsupported instruction
+/// before we even get to executing this code.
+pub fn ensure_cpu_compatibility(force: bool) {
+    if cfg!(near_production) || force {
+        imp::ensure_cpu_compatibility();
+    } else {
+        // Intentionally empty: allow execution without verifying the properties of the CPU.
+    }
+}
+
+mod imp {
+    const NOTE: &str = "confirm hardware requirements at \
+        <https://docs.near.org/docs/develop/node/validator/hardware>.";
+
+    #[cfg(not(target_arch = "x86_64"))]
+    pub fn ensure_cpu_compatibility() {
+        eprintln!("error: processor running a x86_64 instruction set is required.\nnote: {}", NOTE);
+        std::process::exit(1);
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    pub fn ensure_cpu_compatibility() {
+        // SAFETY: this is “just” a CPUID. Some very ancient chips don't support this instruction
+        // and may do anything else with the opcode.
+        //
+        // As an implementation note, we cannot use `is_*_feature_available!` macro from libstd
+        // here, because it will unconditionally return `true` when `cfg(target_feature="...")` is
+        // true.
+        let cpuid01 = unsafe { std::arch::x86_64::__cpuid(0x01) };
+        ensure_cpuid01(cpuid01.ecx, cpuid01.edx);
+        let cpuid0d = unsafe { std::arch::x86_64::__cpuid(0x0d) };
```

Comment:
Per manuals (https://www.felixcloutier.com/x86/cpuid) leaf 0xd varies depending on ECX value. Better set it.
Looks not very readable, maybe better smth like `LEAF_PROCESSOR_EXTENDED_STATE_ENUMERATION`.

Reply:
`__cpuid` always queries the main sub-leaf (i.e. `ecx` is set to 0)

 https://github.com/rust-lang/stdarch/blob/cfba59fccd90b3b52a614120834320f764ab08d1/crates/core_arch/src/x86/cpuid.rs#L90-L96